daml-lf/spec/contract-id.rst

```diff
@@ -0,0 +1,245 @@
+
+Note that local contract IDs correspond to the ID of output contracts
+together with those contracts that have been created and archived in
+the transaction. On the other hand, global contract IDs do not
+reference only IDs of some contracts that have been previously
+persisted on the ledger, but also any arbitrary contract IDs that the
+submitter have referenced in its submission.
+
+The so-called *discriminator freshness condition* holds if the
+discriminators from local contract IDs are distinct from the
+discriminators from global contract IDs.
+
+This ensures that only the discriminators, not the suffix are needed
+to order the contract IDs of created contract w.r.t. other contract
+IDs. 
+
+Contract ID uniqueness
+----------------------
+
+During interpretation local contract IDs are created without suffix.
+It is the responsibility of the ledger implementation to provide for
+each local contract ID a suffix that guarantee global uniqueness of
+the whole contract ID. No other requirement (except the 95 bytes size
+limit) is assumed for those suffices.
+
+This guard against the submitter choosing low-entropy submission seeds. 
+
+The simplest approach consists to suffix all local contract ID with a
+uniquely global transaction ID. Alternatively central commiter ledger
+can completely avoid suffixing by enforcing that same pair (submission
+seed, submission time) is never used more that once, the discriminator
+allocation scheme ensuring in this case the uniqueness of allocated
+discriminators.
+
+Submission time
+---------------
+
+The submission time is used to derive the transaction seed. In
+practice, it should be closed to the original time at which the
+submission was initiated. No particular requirement is made in this
+specification. However a particular ledger implementation can require
+this time to be within a particular time window around the commit
+time. This allows, for instance, the commiter of a central commiter
+ledger to en enforce global uniqueness of the pair (submission seed,
+submission time).
+
+Allocation scheme for discriminators
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+  
+From an initial seed, called in the following *submission seed*, seeds
+for all nodes in the transaction tree are derived as they are
+computed.
+
+We propose the following scheme, inspired by NIST’s HMAC_DRBG
+construction. It assumes a hash-based MAC function HMAC, and defines
+for each big-endian 32 bits integer ``n`` the derived seed::
+
+  deriveSeed(seed, n) := HMAC(seed, n)
+  
+
+Submission seed
+---------------
+
+Any random number with sufficient entropy works as submission seed
+``submission_seed`` (256 bit entropy).
+
+Transaction seed construction
+-----------------------------
+
+From the submission, a so-called *transaction seed* is derived as follows::
+
+  deriveTransactionSeed(submissionSeed, participandId, submissionTime, submitters) :=
+     HMAC(submissionSeed, participantId ∥ submissionTime ∥ nₛ ∥ submitters)
```

Comment:
`participantId` is a ledger string, i.e., of varying length. In the above notation, I do not see that the length of `participantId` is included in the argument to `HMAC`. If it isn't, we may get hash collissions because `submitters` too is a variable-length string. In the implementation I see that the actual derivation adds the length of the participantId to the string. Can you also do so here?

Reply:
Found the comment below about the length prefix.